aten/src/ATen/native/mps/operations/Convolution.mm

```diff
@@ -158,57 +222,77 @@ Tensor _mps_convolution_impl(const Tensor& input_t,
+                                             dilation[0],
+                                             padding[1],
+                                             padding[0],
+                                             memory_format,
+                                             groups);
+                } else {
+                    fill_conv_desc(conv2dDescriptor_,
+                                   stride[1],
+                                   stride[0],
+                                   dilation[1],
+                                   dilation[0],
+                                   padding[1],
+                                   padding[0],
+                                   memory_format,
+                                   groups);
+                }
+            }
 
           MPSGraphTensor* inputTensor = native_mps::mpsGraphRankedPlaceHolder(
               mpsGraph, native_mps::getMPSScalarType(input_t.scalar_type()), inputShape);
           MPSGraphTensor* weightTensor = native_mps::mpsGraphRankedPlaceHolder(mpsGraph, weight_t);
 
-          MPSGraphTensor* biasTensor = nil;
-          if (bias_defined) {
-            biasTensor =
-                native_mps::mpsGraphUnrankedPlaceHolder(mpsGraph, native_mps::getMPSDataType(bias_opt.value()));
-          }
-
-          MPSGraphTensor* outputTensor;
-          if (isDepthwiseConv) {
-            MPSGraphTensor* weightTransposeTensor = [mpsGraph transposeTensor:weightTensor
-                                                                    dimension:-3
-                                                                withDimension:-4
-                                                                         name:nil];
-            outputTensor = [mpsGraph depthwiseConvolution3DWithSourceTensor:inputTensor
-                                                              weightsTensor:weightTransposeTensor
-                                                                 descriptor:depthWiseConv3dDescriptor_
-                                                                       name:nil];
-          } else {
-            outputTensor = [mpsGraph convolution2DWithSourceTensor:inputTensor
-                                                     weightsTensor:weightTensor
-                                                        descriptor:conv2dDescriptor_
-                                                              name:nil];
-          }
+            MPSGraphTensor* biasTensor = nil;
+            MPSGraphTensor* outputTensor;
+            if(is3DConv){
+                outputTensor = [mpsGraph convolution3DWithSourceTensor:inputTensor
+                                                         weightsTensor:weightTensor
+                                                            descriptor:conv3dDescriptor_
+                                                                  name:nil];
+                
+            } else {
+                if (bias_defined) {
```

Comment:
This conditional statement that contains the biasTensor assignment should be outside the conditional block that contains it. Otherwise, `biasTensor` will always be `nil` for 3d convolutions.

Reply:
You can move it right below where you set `MPSGraphTensor* biasTensor = nil;`

Reply:
Thanks! I moved the block accordingly so that it is called regardless of isConv3d or isDepthwiseConv. I amended my tests to include biases and it now works as expected. I created a set of tests at https://gist.github.com/mattiaspaul/b63cd65c9afa4290b316d9297e19ca03 (maybe some could be added to the official test_mps.py at some point)